cypress-io/cypress
# fix: remove working directory from `--spec` glob patterns

- Closes #22272

### User facing changelog


* Fixed an issue where usage of the `--spec` option would result in no specs found if the absolute project path included `glob` pattern characters.

### Additional details


We document the [`--spec`](https://docs.cypress.io/guides/guides/command-line#cypress-run-spec-lt-spec-gt) run mode option as supporting:
* Relative paths
* Absolute paths
* Glob patterns
* Comma-delimited lists of the above

Given this range of support, we normalize the provided value into an absolute path during option processing. This value eventually is given to the FileDateSource for matching against the filesystem using globby. However, the absolute path can contain glob pattern syntax (e.g. `C:\Program Files (x86)\my-project\...`) that is parsed by globby and can cause our lookups to fail. Today, users have little recourse but to move their project into a different directory, which isn't always possible.

My initial thought was to stop converting the option value into an absolute path altogether, but there were test cases featuring relative pathing that made me question the passivity of such a solution in a patch release. I instead updated the logic around our globbing to ensure that the working directory is never present in the patterns we supply to globby. We limit the search space to the current working directory using separate globby options, so it is not needed within the pattern itself. I also added a few unit tests around the updated logic and an additional system test with a project that includes these troublesome characters.

These issues are somewhat related to those I raised around our expected `supportFile` values [here](https://github.com/cypress-io/cypress/issues/22255), namely that it is hard for us to support absolute paths, relative paths, and glob patterns through a single API, especially across our supported architectures. Simplifying these options down to supporting only glob patterns, relative to the project root, would allow us to simplify a lot of internal transforms we're doing and leave us with these potential values:

```js
const supportedPatterns = [
  'cypress/e2e/*.cy.js',                 // standard patterns
  'cypress/e2e/spec.cy.js' ,             // patterns to match single files
  'cypress/e2e/spec \\(1\\).cy.js',      // patterns with user-escaped glob chars
  '**/spec-1.cy.js,**/spec-2.cy.js',     // we could still support multiple patterns if we want to...
  '**/spec-(1|2).cy.js',                 // ...but having defined glob expectations might negate the need
]

const unsupportedPatterns = [
  'cypress\e2e\*.cy.js',      // on windows, users would still use forward-slashes (backwards are reserved for escaping)
  'cypress/e2e/spec (1).js'.  // this will not match against an explicit file including '()', chars must be escaped.
]
```  
  
### Steps to test


* Check out https://github.com/tbiethman/validate-23296 into a directory like: `/Users/foo/my-(projects)/validate-23296`
* Run `yarn cypress:run --spec "**/*.cy.js"`, see "No specs found" error
* Install arch-appropriate binary from this branch [commit](https://github.com/cypress-io/cypress/commit/2cfd4a95c804ad763d5cb037e83f780ad09c44be)
* Run `yarn cypress:run --spec "**/*.cy.js"` again, see resulting success

### How has the user experience changed?


### PR Tasks


- [n/a] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [n/a] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
